### ISSUE TITLE
Improve the `useContext` logic

### ISSUE THREAD

**Issue by WebReflection** [WebReflection/uhooks#2]
Apparently, the current logic is [not so useful or friendly](https://github.com/WebReflection/uland/issues/14), compared to the *React* context counter-part.

**edit** it's also possible *uhhoks* are not responsible for issues, are client works, but SSR doesn't:
https://github.com/WebReflection/uland-ssr/issues/2

**Comment by WebReflection** [comment]
### Background

The *uland-ssr* only specific issue [has been fixed](https://github.com/WebReflection/uland-ssr/pull/3).
The presented use case works now without issues, but it's been stated that *React* works better both synchronously and asynchronously, so here the analysis of the issue.

## Analysis

Given [the following React example](https://codepen.io/WebReflection/pen/vYmBRqz?editors=0010), it is possible to indeed verify that *React* works without issues asynchronously too.

```js
import React, {useContext} from '//cdn.skypack.dev/react';
import ReactDOM from '//cdn.skypack.dev/react-dom';

const Value = React.createContext();

function Display() {
  const value = useContext(Value);
  return <div>The value is {value}.</div>;
}

async function doRender(where, value, delay) {
  console.log('awaiting', value);
  await new Promise($ => setTimeout($, delay));
  ReactDOM.render((
    <Value.Provider value={value}>
      <Display />
    </Value.Provider>
  ), where);
  console.log('rendered', value);
}

doRender(first, "123", 500);
doRender(second, "456", 200);
```

However, there is a subtle, yet very relevant, difference with the originally proposed @jaydenseric issue:

```js
async function doRender(fooValue) {
  // the context is shared, and its most updated value set here
  FooContext.provide(fooValue);

  // a delay is present **in between** the context update and the render
  await new Promise((resolve) => setTimeout(resolve, 1));

  // a render is done with FooComponent using the context
  render(String, () => html`
    <!doctype html>
    <html>${FooComponent()}</html>
  `);
}

// set latest context value to `123`
doRender("123");
// set latest context value to `456` **synchronously**
doRender("456");
```

### When React context is resolved

Given this part of the code:

```js
async function doRender(where, value, delay) {
  await new Promise($ => setTimeout($, delay));
  ReactDOM.render((
    <Value.Provider value={value}>
      <Display />
    </Value.Provider>
  ), where);
}
```

The resulting code, after transformation, would be:

```js
var _ref = /*#__PURE__*/ React.createElement(Display, null);

async function doRender(where, value, delay) {
  // the delay happens **before** the value is passed around
  await new Promise(($) => setTimeout($, delay));
  ReactDOM.render(
    /*#__PURE__*/ React.createElement(
      Value.Provider,
      {
        // here the value is the expected one **when** render happens
        value: value
      },
      _ref
    ),
    where
  );
}
```

## The uland and uhooks case

Given [the following code](https://codepen.io/WebReflection/pen/NWjKMwN?editors=0010), the result, or better, the value is passed along same way the *React* counter example would do.

```js
import {render} from 'https://cdn.skypack.dev/uland/jsx';

function Display({value}) {
  return <div>The value is {value}.</div>;
}

async function doRender(where, value, delay) {
  console.log('awaiting', value);
  await new Promise($ => setTimeout($, delay));
  render(where, <Display value={value} />);
  console.log('rendered', value);
}

doRender(first, "123", 500);
doRender(second, "456", 200);
```
Not only this is easier to reason about, as `value` is propagated down the road at the right time, and not retrieved too late in the stack, using `context.provide(value)` **outside** a component, would not be allowed, as `provide`, on the client, is meant to update all components that used such context.

### Client VS Server

On the SSR version of *uhooks*, namely [uhooks-nofx](https://github.com/WebReflection/uhooks-nofx), nothing react to state or context changes, as the rendered component cannot really get updated on the fly (it's streamed content, after all), so it might be possible to use `context.provide(value)` at any time, even outside components.

However, the right way way to do so would be by providing the value **at the right time**, and not before.

[Example](https://codepen.io/WebReflection/pen/YzVKLOo?editors=0010)

```js
import {
  Component, render, html,
  createContext, useContext
} from 'https://cdn.skypack.dev/uland-ssr';

const Value = createContext();

const Display = Component(() => {
  const value = useContext(Value);
  return html`<div>The value is ${value}.</div>`;
});

async function doRender(where, value, delay) {
  console.log('awaiting', value);
  await new Promise($ => setTimeout($, delay));
  // this is where context.provide makes sense, not before
  Value.provide(value);
  const result = render(String, html`${Display()}`);
  where.innerHTML = result;
  console.log('rendered', value);
}

doRender(first, "123", 500);
doRender(second, "456", 200);
```

## What to improve

Because the children, even if pre-invoked, will be resolved only on render, hence their call stack will access the right context at the right time only there, and because updating a context with a value *before* asynchronous operations and render, is an anti pattern, and there's no logical solution to it, the only thing to improve here is being able to normalize the usage of the context so that it's possible to call it **even outside a module**, updating callbacks only when the rendering stack is being resolved, but not outside components.

Everything else **works by design**, the issue was fixed in **uland-ssr** side, hence beside guaranteeing feature-parity between **uhooks** and **uhooks-nofx**, when it comes to context, there's not much else to improve here.

**Comment by jaydenseric** [comment]
Thanks for putting time and effort into these issues.

[Here](https://github.com/WebReflection/uland/issues/14#issuecomment-868406827) when I was using `await new Promise($ => setTimeout($, delay));` after setting the context value but before rendering, it was just to simulate a similar result that would happen if the delay was happening inside render (hence the `// Realistically, a delay would happen inside rendering of complex components.` comment), which I couldn't quickly come up with an example for but definitely exist. For example, if doing two sync renders at the same time in parallel, sometimes there can be expensive sync work in the renders that makes it a race condition which progresses faster than the other. It could be processing a massive array, doing filters on a canvas, etc.

I suspected there would be a clean reproduction using the async hooks to delay one render more than another, but I just didn't know enough about the API yet to write it.

Regardless, there is a more common situation to consider that highlights problems with having a non-contextual context value design, touched upon here: https://github.com/WebReflection/uland/issues/14#issuecomment-867692275 .

Imagine you are making a GraphQL playground component (`GraphQLPlayground`), that creates a memoized GraphQL client instance constructed using the URI input by the user, via a text input in the `GraphQLPlayground` component. At this level, the `GraphQLPlaygroundContext` is provided, so that composable child components (e.g. `GraphQLPlaygroundDocExplorer`) can be dropped in to add functionality, without any prop drilling of the GraphQL client instance. Now imagine that you have a page that has two of these `GraphQLPlayground` components mounted, with different URIs specified in props as the initial URI. This is the sort of thing context is intended for; a contextual value that only applies in the current branch of the render tree to avoid deep prop drilling. With the current uland context system, the second `GraphQLPlayground` component to render on the page will overwrite the context value that was provided in the first.

**Comment by jaydenseric** [comment]
Another situation to consider is repeat waterfall rendering in SSR, you can learn more about the technique here:

https://github.com/jaydenseric/react-waterfall-render#function-waterfallrender

A significant portion of Apollo apps use a similar technique.

Even if the renders are sync, in such a situation with repeat renders for each request with varying loading delays between repetitions, with multiple different user requests being server side rendered at once, we will see corruption of context values and therefore the rendered HTML served to the users. At unpredictable moments the wrong data will get served to the wrong user.

**Comment by WebReflection** [comment]
Asynchronous components are "*not a thing*" though, the stack will be invoked synchronously in *React*, but it could have asycnhronous hooks inside, but that's up to you to decide if you want to await these (SSR) or update the component on [fetch](https://gist.github.com/WebReflection/dea6cd9ab50e61ec14202a416cd6d121) or anything else (Client).

That being said, if you provide the latest context value *before* the asynchronous part happens, there is no way to prevent the shared context to update its internal value before the rest of the callback is executed.

If you need different contexts per request, then there should be no issue, but having a provide before async invokes and then render, is not what React does, and it's not what any JS can do, without specialized tools around.

The waterfall approach renders whatever it is **right away**, then it instruments the hydration for the client, but this is not strictly what this library does, it might be eventually implemented in **uhooks-nofx** but for React solutions already available out there, my answer would always be: "*if you're looking for a 1:1 React solution, use React, not this library*".

For the time being I will avoid throwing when no component is using context, even if I think that's an anti pattern already, while a context that provides for components make more sense.

Arguably, I am not sure I should even fix that case, but the change is minimal, so it might make sense, although without errors it's harder to realize `useContext` was called outside a component, so I am back to my initial doubt: I don't think there's much I can do here, and the logic around shared context cannot be fixed in the asynchronous case, if the render happens after, but I think I'm repeating myself.

React doesn't have this issue because the context is **always at render time** and never before, so *maybe* I could improve the `uland/jsx` export with a similar `Provider, Consumer` logic, but at the end of the day, that will simply be exact same as moving the context.provide call right before rendering, never sooner.

To recap:

  * if you provide a new value on the client, components that previously used that context will be updated
  * if you provide some value SSR, be sure components depending on that context will point at the right value, so provide the value only before invoking components/render

In short, we have a classic race condition by shared reference, and this problem is not new here, it's a well known one, and the timing is all it matters.

If you have ideas how to solve that though, I'll listen.

P.S. `uland/jsx` is *uland* with `React.createElement` and `React.Fragment` facades to make JSX default Babel transform works with *uland* too, nothing more, nothing less.

**Comment by WebReflection** [comment]
After thinking about it, filtering useContext doesn’t really help, or fix, anything around this async issue, so I think no action here is needed. The async helper use case is interesting for SSR though, but that’s going to be specific for uhooks-nofx, and it won’t solve anyway the usage of provide before async operations.
If I could get a real world example that works in React but not in uland, I might be able to help out fixing that, but the context via jsx looks like something to feature in jsx2tag instead.
